### PR TITLE
Parallel io additions

### PR DESCRIPTION
**Pull request by oHunewald**

### CODE REVIEW

File: src/satellites.jl [hunk 1]
```diff
@@ -519,6 +561,24 @@ function generateIO(filePath, md, nWorkers, generateFiles=true, printLevel=0, sa
     end
 end
 
+function generateIO(filePath, fn::String, nWorkers, generateFiles=true, printLevel=0, saveIndices=false)
```
Comment (laurentheirendt) [inline src/satellites.jl:562]: Can you add here the doc?

File: src/io/input.jl [hunk 2]
```diff
@@ -69,40 +69,102 @@ function readFlowFrame(filename::String)
...
+    # Columns ca be selected by an array of indicies:
+    # loadData(idx, "input-$idx.jls", [3:6;9:11]) <- this will concatenate ranges into arrays
+    # Please note that all optional arguments are by default "false"
+    if type == "fcs"
+        @time @sync for (idx, pid) in enumerate(workers())
+            @async R[idx] = fetch(@spawnat pid loadFCSData(idx, "input-$idx.jls", panel, method,
+                                cofactor,reduce, sort, transform))
+        end
+    else
+        @info "File Type not yet supported!"
```
Comment (laurentheirendt) [inline src/io/input.jl]: this should be an `@error`

File: src/io/input.jl [hunk 3]
```diff
@@ -69,40 +69,102 @@ function readFlowFrame(filename::String)
...
+    loadData(dataPath, data, nWorkers; panel=Nothing(), 
+            type = "fcs", method = "asinh", cofactor = 5,
+            reduce = false, sort = false, transform = false)
+
+This function is of 2 parts. Part 1: Generates the temporary binaray files to be loaded by the
+    workers. The Input data will be equally divided into n parts according to the number of workers.
+    Part2: each worker loads independently its own data-package in parallel and returns
+
+# Arguments:
+- `dataPath`: worker index
```
Comment (laurentheirendt) [inline src/io/input.jl]: worker index?

File: src/io/input.jl [hunk 4]
```diff
@@ -69,40 +69,102 @@ function readFlowFrame(filename::String)
 end
 
 """
-    loadData(idx, fn, md,panel; method = "asinh", cofactor = 5,
-            reduce = true, sort = true)
+    loadData(dataPath, data, nWorkers; panel=Nothing(),
```
Comment (laurentheirendt) [inline src/io/input.jl]: is it worth typing `data`?

File: src/io/input.jl [hunk 5]
```diff
@@ -69,40 +69,102 @@ function readFlowFrame(filename::String)
...
+
+This function is of 2 parts. Part 1: Generates the temporary binaray files to be loaded by the
+    workers. The Input data will be equally divided into n parts according to the number of workers.
+    Part2: each worker loads independently its own data-package in parallel and returns
+
+# Arguments:
+- `dataPath`: worker index
+- `data`: single filename::String or a metadata::DataFrame with a column sample_name
+- `panel`: Panel table with a column for Lineage Markers and one for Functional Markers,
+    or Array::{Int} used as column indicies, default: Nothing()
```
Comment (laurentheirendt) [inline src/io/input.jl]: indices

Comment (laurentheirendt) [inline src/io/input.jl]: why `Nothing()`?

File: src/io/input.jl [hunk 6]
```diff
@@ -69,40 +69,102 @@ function readFlowFrame(filename::String)
...
+    R =  Vector{Any}(undef,nWorkers)
+
+    # Load the data by each worker
+    # Without panel file, all columns are loaded:
+    # loadData(idx, "input-$idx.jls")
+    # Columns ca be selected by an array of indicies:
+    # loadData(idx, "input-$idx.jls", [3:6;9:11]) <- this will concatenate ranges into arrays
+    # Please note that all optional arguments are by default "false"
+    if type == "fcs"
+        @time @sync for (idx, pid) in enumerate(workers())
```
Comment (laurentheirendt) [inline src/io/input.jl]: there is no need to benchmark here. I think you can remove `@time`.

File: src/io/input.jl [hunk 7]
```diff
@@ -69,40 +69,102 @@ function readFlowFrame(filename::String)
...
+- `sort`: Sort columns by name to make sure the order when concatinating the dataframes, optional, default: false
+- `transform`: Boolean to indicate if the data will be transformed according to method, default: false
+"""
+function loadData(dataPath, data, nWorkers; panel=Nothing(), 
+                type = "fcs", method = "asinh", cofactor = 5,
+                reduce = false, sort = false, transform = false)
+
+    # Split the data according to the number of worker as temp binary file
+    # md can be a metadata file or a single file_name:
+    # generateIO(dataPath, <filename>, nWorkers, true, 1, true)
```
Comment (laurentheirendt) [inline src/io/input.jl]: I think this line can be removed.

File: src/io/input.jl [hunk 8]
```diff
@@ -69,40 +69,102 @@ function readFlowFrame(filename::String)
...
+    xRange = generateIO(dataPath, data, nWorkers, true, 1, true)
```
Comment (laurentheirendt) [inline src/io/input.jl:98]: What does `xRange` stand for?

File: src/io/input.jl [hunk 9]
```diff
@@ -69,40 +69,102 @@ function readFlowFrame(filename::String)
...
-- `panel`: Panel table with a column for Lineage Markers and one for Functional Markers
+- `panel`: Panel table with a column for Lineage Markers and one for Functional Markers, 
+    or Array::{Int} used as column indicies
 - `method`: transformation method, default arcsinh, optional
 - `cofactor`: Cofactor for transformation, default 5, optional
 - `reduce`: Selected only columns which are defined by lineage and functional, optional,
     default: true. If false the check for any none columns to be removed (none columns can appear
     after concatenating FCS files as well as parameter like: time, event length)
 - `sort`: Sort columns by name to make sure the order when concatinating the dataframes, optional, default: true
+- `transform`: Boolean to indicate if the data will be transformed according to method
```
Comment (laurentheirendt) [inline src/io/input.jl:137]: I think there should be a default value for `transform`

File: src/satellites.jl [hunk 10]
```diff
@@ -231,6 +231,48 @@ function getTotalSize(loc, md, printLevel=0)
     return totalSize, inSize, runSum
 end
 
+
+function getTotalSize(loc, fn::String, printLevel=0)
```
Comment (laurentheirendt) [inline src/satellites.jl]: can you add a documentation here?

Comment (laurentheirendt) [reply src/satellites.jl]: is there any way that we could combine both overloaded `getTotalSize` functions?

File: src/satellites.jl [hunk 11]
```diff
@@ -543,4 +603,23 @@ function rmFile(fileName, printLevel = 1)
             printstyled("(file $fileName does not exist - skipping).\n", color=:red)
         end
     end
+end
+
+# Statically split range [1,N] into equal sized chunks for np processors
+function splitrange(N::Int, np::Int)
```
Comment (laurentheirendt) [inline src/satellites.jl]: can you add a more extensive documentation here?

File: test/testLoadData.jl [hunk 12]
```diff
@@ -0,0 +1,97 @@
+using GigaSOM, DataFrames, XLSX, CSV, Test, Random, Distributed, SHA, JSON
```
Comment (laurentheirendt) [inline test/testLoadData.jl]: can this test be combined with the `testIO.jl` file?

File: test/testSingleFileSplitting.jl [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+df = GigaSOM.transformData(df, "asinh", 5)
+df_firstLine = df[1,:]
+
+nWorkers = 2
+addprocs(nWorkers, topology=:master_worker)
+@everywhere using GigaSOM, FCSFiles
+
+R, = loadData(dataPath, fn, nWorkers)
+
+@testset "check ranges" begin
```
Comment (laurentheirendt) [inline test/testSingleFileSplitting.jl]: I think these test sets can be combined

File: test/testTrainingOuputEquality.jl [hunk 14]
```diff
@@ -15,21 +15,15 @@ nWorkers = 2
...
 @everywhere using GigaSOM, FCSFiles
 
-generateIO(dataPath, md, nWorkers, true, 1, true)
-
-R =  Vector{Any}(undef,nWorkers)
-
-@sync for (idx, pid) in enumerate(workers())
-    @async R[idx] = fetch(@spawnat pid loadData(idx, "input-$idx.jls", md, panel))
-end
+R, _ = loadData(dataPath, md, nWorkers, panel=panel, reduce=true, transform=true)
```
Comment (laurentheirendt) [inline test/testTrainingOuputEquality.jl]: I think `_` is not necessary.

**Review by laurentheirendt — CHANGES_REQUESTED** [review]
first round of review

File: src/io/input.jl (hunk 4 above)
Comment (oHunewald) [reply src/io/input.jl]: yes, data could be a single data file or a meta-data file

File: src/io/input.jl (hunk 5 above)
Comment (oHunewald) [reply src/io/input.jl]: according to Julia documentation this is used for NULL, and like this I can check if the parameter is set or not

File: src/io/input.jl (hunk 8 above)
Comment (oHunewald) [reply src/io/input.jl:98]: thats for the data ranges in each worker data file (needed for later processing)

File: src/io/input.jl (hunk 9 above)
Comment (oHunewald) [reply src/io/input.jl:137]: loadFCS() is an internal function and transform is set in the caller function before. It is only passed through

File: src/satellites.jl (hunk 10 above)
Comment (oHunewald) [reply src/satellites.jl]: yes, now combined

File: src/satellites.jl (hunk 11 above)
Comment (oHunewald) [reply src/satellites.jl]: I would appreciate if you could do this, don't know exactly what you did

File: test/testLoadData.jl (hunk 12 above)
Comment (oHunewald) [reply test/testLoadData.jl]: I would prefer to keep this test separated as testFileIO is using the "classic" approach and will be depreciated in a later stage.

Comment (laurentheirendt) [reply test/testLoadData.jl]: ok with me

File: src/satellites.jl (hunk 11 above)
Comment (laurentheirendt) [reply src/satellites.jl]: I didn't write the `splitrange` function. Can you combine it with the `splitting` function?

**Review by laurentheirendt — APPROVED** [review]

**Comment by laurentheirendt** [comment]
Thanks @oHunewald 👍

**Pull request merged**